File: frontend/src/services/statistics.ts
/**
 * 统计服务
 */
import apiClient from './api'
import { API_ENDPOINTS } from '../config/api'

export interface TaskStatistics {
  total: number
  pending: number
  confirmed: number
  detail_submitted: number
  dispatched: number
  in_progress: number
  completed: number
  rejected: number
  cancelled: number
  completion_rate: number
}

export interface OpportunityStatistics {
  total: number
  created: number
  in_progress: number
  lost: number
  won: number
  conversion_rate: number
  loss_rate: number
}

export interface WorkOrderStatistics {
  total: number
  pending: number
  accepted: number
  in_progress: number
  completed: number
  cancelled: number
  completion_rate: number
}

export interface VisitLogStatistics {
  total: number
  with_opportunity: number
  without_opportunity: number
  opportunity_rate: number
}

export interface ReviewStatistics {
  total: number
}

export interface MemberWorkload {
  member_id: number
  member_name: string
  work_order_count: number
  visit_log_count: number
  completed_work_orders: number
}

export interface DashboardStatistics {
  tasks: TaskStatistics
  opportunities: OpportunityStatistics
  work_orders: WorkOrderStatistics
  visit_logs: VisitLogStatistics
  reviews: ReviewStatistics
  member_workloads: MemberWorkload[]
}

export interface TimeRangeStatistics {
  date: string
  task_count: number
  work_order_count: number
  visit_log_count: number
  opportunity_count: number
}

export interface SalesUnitStatistics {
  sales_unit: string
  task_count: number
  work_order_count: number
  opportunity_count: number
  conversion_rate: number
}

export interface SalesUnitPerformanceStatistics {
  sales_unit: string
  appointments_made: number  // 已预约（工单数量）
  visits_completed: number  // 已拜访（拜访日志数量）
  effective_appointment_rate: number  // 有效预约率
  has_decision_authority: number  // 有拜访对象权限（建议权或决策权）
  effective_visit_rate: number  // 有效拜访率
  lead_count: number  // 线索数量
  opportunity_count: number  // 商机数量
  lead_mining_rate: number  // 线索挖掘率
  lead_conversion_rate: number  // 线索转化率
}

export interface RequirementDirectionStatistics {
  direction: string
  count: number
}

export interface RequirementDirectionGroupStatistics {
  category: string
  directions: RequirementDirectionStatistics[]
}

export interface MemberDetailStatistics {
  member_id: number
  member_name: string
  appointments_made: number
  visits_completed: number
  effective_appointment_rate: number
  has_decision_authority: number
  effective_visit_rate: number
  lead_count: number
  opportunity_count: number
  lead_mining_rate: number
  lead_conversion_rate: number
}

export interface SalesUnitPerformanceResponse {
  statistics: SalesUnitPerformanceStatistics[]
  requirement_directions: RequirementDirectionGroupStatistics[]
  member_details?: MemberDetailStatistics[]
}

export interface OpportunityConvertedAmountStatistics {
  member_id?: number
  member_name?: string
  group_id?: number
  group_name?: string
  converted_count: number
  total_amount: number
}

export const statisticsService = {
  /**
   * 获取工作台统计数据
   */
  getDashboardStatistics: async (params?: {
    group_id?: number
    start_date?: string
    end_date?: string
  }): Promise<DashboardStatistics> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_DASHBOARD, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取任务统计
   */
  getTaskStatistics: async (params?: {
    start_date?: string
    end_date?: string
  }): Promise<TaskStatistics> => {
    const queryParams: any = {}
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_TASKS, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取商机统计
   */
  getOpportunityStatistics: async (params?: {
    group_id?: number
    start_date?: string
    end_date?: string
  }): Promise<OpportunityStatistics> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_OPPORTUNITIES, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取工单统计
   */
  getWorkOrderStatistics: async (params?: {
    group_id?: number
    start_date?: string
    end_date?: string
  }): Promise<WorkOrderStatistics> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_WORK_ORDERS, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取成员工作量统计
   */
  getMemberWorkloads: async (params?: {
    limit?: number
    group_id?: number
    start_date?: string
    end_date?: string
  }): Promise<MemberWorkload[]> => {
    const queryParams: any = {}
    if (params?.limit) queryParams.limit = params.limit
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_MEMBER_WORKLOAD, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取时间段统计
   */
  getTimeRangeStatistics: async (params?: {
    start_date?: string
    end_date?: string
    group_by?: 'day' | 'week' | 'month'
    group_id?: number
  }): Promise<TimeRangeStatistics[]> => {
    return apiClient.get(API_ENDPOINTS.STATISTICS_TIME_RANGE, {
      params
    })
  },

  /**
   * 获取销售单位统计
   */
  getSalesUnitStatistics: async (params?: {
    group_id?: number
    start_date?: string
    end_date?: string
  }): Promise<SalesUnitStatistics[]> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_SALES_UNITS, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取销售单位绩效统计
   */
  getSalesUnitPerformanceStatistics: async (params?: {
    group_id?: number
    include_member_details?: boolean
    start_date?: string
    end_date?: string
  }): Promise<SalesUnitPerformanceResponse> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.include_member_details) queryParams.include_member_details = params.include_member_details
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_SALES_UNITS_PERFORMANCE, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },

  /**
   * 获取转订商机总金额统计
   */
  getOpportunityConvertedAmountStatistics: async (params?: {
    group_id?: number
    include_member_details?: boolean
    start_date?: string
    end_date?: string
  }): Promise<OpportunityConvertedAmountStatistics[]> => {
    const queryParams: any = {}
    if (params?.group_id) queryParams.group_id = params.group_id
    if (params?.include_member_details) queryParams.include_member_details = params.include_member_details
    if (params?.start_date) queryParams.start_date = params.start_date
    if (params?.end_date) queryParams.end_date = params.end_date
    return apiClient.get(API_ENDPOINTS.STATISTICS_OPPORTUNITY_CONVERTED_AMOUNT, {
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined
    })
  },
}

File: backend/app/schemas/statistics.py
"""
统计数据相关的 Pydantic 模式
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Tuple
from datetime import date, datetime


class TaskStatistics(BaseModel):
    """任务统计"""
    total: int = 0
    pending: int = 0
    confirmed: int = 0
    detail_submitted: int = 0
    dispatched: int = 0
    in_progress: int = 0
    completed: int = 0
    rejected: int = 0
    cancelled: int = 0
    completion_rate: float = 0.0  # 完成率


class OpportunityStatistics(BaseModel):
    """商机统计"""
    total: int = 0
    created: int = 0
    in_progress: int = 0
    lost: int = 0
    won: int = 0
    conversion_rate: float = 0.0  # 转化率（转定/总数）
    loss_rate: float = 0.0  # 流失率（流失/总数）


class WorkOrderStatistics(BaseModel):
    """工单统计"""
    total: int = 0
    pending: int = 0
    accepted: int = 0
    in_progress: int = 0
    completed: int = 0
    cancelled: int = 0
    completion_rate: float = 0.0  # 完成率


class VisitLogStatistics(BaseModel):
    """拜访日志统计"""
    total: int = 0
    with_opportunity: int = 0
    without_opportunity: int = 0
    opportunity_rate: float = 0.0  # 商机率


class ReviewStatistics(BaseModel):
    """复盘统计"""
    total: int = 0


class MemberWorkload(BaseModel):
    """成员工作量"""
    member_id: int
    member_name: str
    work_order_count: int = 0
    visit_log_count: int = 0
    completed_work_orders: int = 0


class DashboardStatistics(BaseModel):
    """工作台统计数据"""
    tasks: TaskStatistics
    opportunities: OpportunityStatistics
    work_orders: WorkOrderStatistics
    visit_logs: VisitLogStatistics
    reviews: ReviewStatistics
    member_workloads: List[MemberWorkload] = []


class TimeRangeStatistics(BaseModel):
    """时间段统计"""
    date: date
    task_count: int = 0
    work_order_count: int = 0
    visit_log_count: int = 0
    opportunity_count: int = 0


class SalesUnitStatistics(BaseModel):
    """销售单位统计"""
    sales_unit: str
    task_count: int = 0
    work_order_count: int = 0
    opportunity_count: int = 0
    conversion_rate: float = 0.0


class SalesUnitPerformanceStatistics(BaseModel):
    """销售单位绩效统计（按客户来源分组）"""
    sales_unit: str  # 客户来源（销售单位）
    appointments_made: int = 0  # 已预约（工单数量）
    visits_completed: int = 0  # 已拜访（拜访日志数量）
    effective_appointment_rate: float = 0.0  # 有效预约率 = 已拜访 / 已预约
    has_decision_authority: int = 0  # 有拜访对象权限（建议权或决策权）的拜访日志数量
    effective_visit_rate: float = 0.0  # 有效拜访率 = 有决策/建议权 / 已拜访
    lead_count: int = 0  # 线索数量
    opportunity_count: int = 0  # 商机数量
    lead_mining_rate: float = 0.0  # 线索挖掘率 = 线索数量 / 有决策/建议权
    lead_conversion_rate: float = 0.0  # 线索转化率 = 商机数量 / 线索数量


class RequirementDirectionStatistics(BaseModel):
    """需求方向统计"""
    direction: str  # 需求方向名称
    count: int = 0  # 数量


class RequirementDirectionGroupStatistics(BaseModel):
    """需求方向分组统计"""
    category: str  # 分类：算力、模型、应用
    directions: List[RequirementDirectionStatistics] = []  # 该分类下的需求方向统计列表


class MemberDetailStatistics(BaseModel):
    """成员明细统计"""
    member_id: int
    member_name: str
    appointments_made: int = 0  # 已预约（工单数量）
    visits_completed: int = 0  # 已拜访（拜访日志数量）
    effective_appointment_rate: float = 0.0  # 有效预约率
    has_decision_authority: int = 0  # 有拜访对象权限（建议权或决策权）
    effective_visit_rate: float = 0.0  # 有效拜访率
    lead_count: int = 0  # 线索数量
    opportunity_count: int = 0  # 商机数量
    lead_mining_rate: float = 0.0  # 线索挖掘率
    lead_conversion_rate: float = 0.0  # 线索转化率


class SalesUnitPerformanceResponse(BaseModel):
    """销售单位绩效统计响应"""
    statistics: List[SalesUnitPerformanceStatistics] = []  # 按销售单位分组的统计
    requirement_directions: List[RequirementDirectionGroupStatistics] = []  # 需求方向统计（按分类分组）
    member_details: Optional[List[MemberDetailStatistics]] = None  # 成员明细统计（仅组长可用）


class OpportunityConvertedAmountStatistics(BaseModel):
    """转订商机总金额统计"""
    member_id: Optional[int] = None  # 成员ID（仅成员明细时使用）
    member_name: Optional[str] = None  # 成员名称（仅成员明细时使用）
    group_id: Optional[int] = None  # 分组ID（仅总管查看组明细时使用）
    group_name: Optional[str] = None  # 分组名称（仅总管查看组明细时使用）
    converted_count: int = 0  # 转订商机数量
    total_amount: float = 0.0  # 转订商机总金额（元）